File: testdata/fixtures/scala/CWE-79.scala
// CWE-79: XSS via Play Html — untrusted input reaches raw HTML without escaping
import play.api.templates.Html
import javax.servlet.http.HttpServletRequest

object XssTest {
  def vuln(request: HttpServletRequest) = {
    val user = request.getParameter("name")
    // ruleid: scala-xss
    Html(user)
    // ruleid: scala-xss
    Html.apply(user)
    val data = request.getHeader("X-Data")
    // ruleid: scala-xss
    Html(data)
    // propagation
    val propagated = user
    // ruleid: scala-xss
    Html(propagated)
  }
  def safe() = {
    // ok: scala-xss
    Html("fixed string")
    // ok: scala-xss
    Html.apply("hello")
    // ok: scala-xss
    val safe = "constant"
    Html(safe)
  }
}

File: testdata/fixtures/ruby/CWE-79.rb
# CWE-79 Cross-Site Scripting (XSS) in Ruby / Rails
# Rendering user input without escaping via raw, html_safe or render injects
# arbitrary HTML/JS into the response.

def show
  name = params[:name]
  q = params.get("q")
  ck = cookies[:x]
  body = request.body.read

  # ruleid: ruby-xss
  name.html_safe

  # ruleid: ruby-xss
  raw(q)

  # ruleid: ruby-xss
  render ck

  # ruleid: ruby-xss
  raw(body)

  # ok: ruby-xss
  render html: ERB::Util.html_escape("fixed")

  # ok: ruby-xss
  raw("fixed string")

  # ok: ruby-xss
  render html: "hello world".html_safe
end
